extensions/vault/runtime/src/main/java/io/quarkus/vault/config/VaultConfigSource.java

```diff
@@ -0,0 +1,264 @@
+package io.quarkus.vault.config;
+
+import static io.quarkus.vault.CredentialProvider.PASSWORD_PROPERTY_NAME;
+import static io.quarkus.vault.config.VaultAuthType.K8S;
+import static io.quarkus.vault.config.VaultCacheEntry.tryReturnLastKnownValue;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_CACHE_PERIOD;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_CONNECT_TIMEOUT;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_K8S_TOKEN_PATH;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_KV_SECRET_ENGINE_PATH;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_READ_TIMEOUT;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_RENEW_GRACE_PERIOD;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.DEFAULT_TLS_SKIP_VERIFY;
+import static io.quarkus.vault.config.VaultServerRuntimeConfig.KV_SECRET_ENGINE_VERSION_V1;
+import static io.quarkus.vault.internal.LogConfidentialityLevel.MEDIUM;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.Integer.parseInt;
+import static java.util.stream.Collectors.toMap;
+
+import java.time.Duration;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.configuration.DurationConverter;
+import io.quarkus.vault.internal.LogConfidentialityLevel;
+import io.quarkus.vault.internal.VaultManager;
+
+public class VaultConfigSource implements ConfigSource {
+
+    private static final Logger log = Logger.getLogger(VaultConfigSource.class);
+
+    private static final String PROPERTY_PREFIX = "quarkus.vault.";
+    private static final Pattern adCredsPattern = Pattern.compile("^quarkus.vault.ad.([^.]+).password$");
+    public static final Pattern CREDENTIAL_PATTERN = Pattern.compile("^quarkus\\.vault\\.credential\\.([^.]+)\\.");
```

Comment:
It's a bit unfortunate that we can't leverage our  config framework a bit better. @dmlloyd do you have any idea on how we could improve this?

BTW, just to be sure: you don't have warnings about unused properties in the log when you define this type of property?

Reply:
> BTW, just to be sure: you don't have warnings about unused properties in the log when you define this type of property?

all properties loaded by hand are also represented as valid quarkus properties in the vault config classes: VaultServerRuntimeConfig, CredentialProviderConfig.
I only get warnings when those classes and the adhoc config loading in the vault config source is not well aligned. I checked, I do not have any at this time.